### DISCUSSION TITLE
Predicting with dummy variable

### DISCUSSION THREAD

**Discussion by orrortega** [inlabru-org/inlabru#186] [Q&A]
Hi all,

I am working with a spatiotemporal model with 16 time periods and (4 years) with four seasons per year. We have included the quarters (Q1, Q2, Q3 and Q4) as a fixed effect in the model, and then the temporal effect for the continuity of these 16 periods.

Everything looks ok, and the posterior distributions look correct. However, when we try to plot the predictions, we cannot see much different between quarters (something that in the summary of the model is clear).

We have tried different ways, with the prediction and looks like predicting 4 times (each time only the Quarter that we want to predict) is showing better results, but we don't know if we are correct.  

This is the chunk of code that we are using for the model and the predictions:

```r
cmp1 <- count ~
  field(main = coordinates,
        model = matern,
        group = timePeriod,
        group_mapper = bru_mapper_index(ntimePeriod),
        control.group = list(model = model_type)
        ) +
  Intercept(1) +
  Q2 + Q3 + Q4 +
  myoffset(log(data_sp$users_nz), model = "offset")

bru_model1 = bru(cmp1,
                 family = "zeroinflatedpoisson1",
                 data = data_sp,
                 options = list(verbose = TRUE))

#### with this one is hard to see any difference ####
pred1 <-
  predict(bru_model1,
          pxl,
          formula = ~ exp(Intercept + field +
                            Q2 + Q3 + Q4),
          n.samples = n_samples) 
########

#### predict only for quarter 2
predict(bru_model1,
          pxl,
          formula = ~ exp(Intercept + field +
                            Q2),
          n.samples = n_samples)
######
```
I hope this is clear.

Thanks so much

Oscar

**Comment by finnlindgren** [comment]
You skipped the most relevant part; how you create the prediction data `pxl`!  It needs to have the combinations of locations (`coordinates`) and `timePeriod`, `Q2`, `Q3`, and `Q4` values you want for the predictions.

**Reply by orrortega** [reply]
Yes, sorry
```r
pxl <- pixels(mesh1, mask=ring_inland1000m_12NM)
pxl <- cprod(data.frame(timePeriod = seq_len(ntimePeriod)),
             pxl)

# timePeriod is the sequence from 1 to 16
```

**Reply by finnlindgren** [reply]
Then Q2, Q3, and Q4 have no values, so you'll just get Q1 predictions, I think.

**Reply by finnlindgren** [reply]
PS. The github discussion system uses markdown, so wrap your code blocks in ` ```r `  and ` ``` ` for nicer typesetting. I edited your posts to use that.

**Reply by orrortega** [reply]
But when we plot the predictions for Q2 using only that fixed effect, the output that changes is only the Quater 2, the rest are the same. The same with Q3 and Q4.

**Reply by finnlindgren** [reply]
Do you have global variables named Q1, Q2, Q3, or Q4? Then it's likely getting some values from there; otherwise I would have expected it to complain about missing inputs.  Do an explicit `pxl$Q1=0`, `pxl$Q2=`, etc to be sure what values are used.

**Reply by finnlindgren** [reply]
It's best to supply explicit values when possible, to avoid getting tangled in R's scoping rules; we could force inlabru to ignore the normal scoping rules, but that would lead to difficulties in other cases.

**Reply by orrortega** [reply]
Yes, we transformed the variable Quarter with 4 levels) to Q2, Q3 and Q4

**Reply by finnlindgren** [reply]
Alternatively, in the prediction expression, use `Intercept + field + Q3_eval(1)` to predict for quarter 3, and similarly for the other quarters.

**Reply by finnlindgren** [reply]
Or no, my last comment doesn't really help here if what you're trying to do is to predict for _all_ quarters at the same time. For that, you have to make sure the pxl contains all the covariate combinations you want to predict for, including all the Q* values.

**Reply by orrortega** [reply]
In fact, the main thing is not if I can do it at the same time, it's if I am doing correctly or there is a better way to do it.

**Reply by finnlindgren** [reply]
Alternatively, redefine the Q* component definitions to use `timePeriod` directly, via the modulus operator `%%`:
```r
cmp1 <- count ~
  field(main = coordinates,
        model = matern,
        group = timePeriod,
        group_mapper = bru_mapper_index(ntimePeriod),
        control.group = list(model = model_type)
        ) +
  Intercept(1) +
  Q2(2 == (timePeriod %% 4)) + Q3(3 == (timePeriod %% 4)) + Q4(0 == (timePeriod %% 4)) +
  myoffset(log(users_nz), model = "const")
```
Then you don't need any Q* dummy covariates at all; it will be computed directly from the `timePeriod` values, when needed.

**Reply by finnlindgren** [reply]
With this last method, your current `predict()` code should work if you always include `+Q2+Q3+Q4` in the formula, without any modification of `pxl`.

**Reply by orrortega** [reply]
Ah, ok thanks! 

But, where is the difference between 

```
cmp1 <- count ~
  field(main = coordinates,
        model = matern,
        group = timePeriod,
        group_mapper = bru_mapper_index(ntimePeriod),
        control.group = list(model = model_type)
        ) +
  Intercept(1) +
  Q2(2 == (timePeriod %% 4)) + Q3(3 == (timePeriod %% 4)) + Q4(0 == (timePeriod %% 4)) +
  myoffset(log(users_nz), model = "const")
```

and 

```
cmp1 <- count ~
  field(main = coordinates,
        model = matern,
        group = timePeriod,
        group_mapper = bru_mapper_index(ntimePeriod),
        control.group = list(model = model_type)
        ) +
  Intercept(1) +
  Q2 + Q3 + Q4 + 
  myoffset(log(users_nz), model = "const")
```
when Q2, Q3 and Q4 are different covariates with 1/0 value depending of the quarter in the timeseries

**Reply by finnlindgren** [reply]
The difference is that in the second version, you have to create the Q2, Q3, Q4 values when doing prediction. With the `timePeriod %% 4` approach, you only need to create the `timePeriod` values.

**Reply by finnlindgren** [reply]
Btw, the group_mapper should become the needed mapper automatically, at least in the latest inlabru version, so I think you can skip that argument, and/or use `ngroup=ntimePeriod` instead.

**Comment by finnlindgren** [comment]
Also, your `myoffset` component may be a problem in some situations, since it ignores the input data; you probably should give it as an `E=` argument to `like() ` instead (the "exposure" parameter of the INLA poisson models).

But if you really _do_ want to be able to include `myoffset` in predictions, use `myoffset(log(users_nz), model="const")` instead ("offset" is an alias of "const"; "const" better describes what such components are)
